### PR TITLE
tiledbsoma 1.12.0 pre-check

### PR DESCRIPTION
**Pull request by johnkerl**
Following our [established procedure](https://github.com/single-cell-data/TileDB-SOMA/wiki/Branches-and-releases)

Significant dialog is on https://github.com/single-cell-data/TileDB-SOMA/pull/2692

### CODE REVIEW

**Review by jdblischak — CHANGES_REQUESTED** [review]
Bump tiledb pin to 2.24. This will result in a conda solver error unless the new binaries with libgoogle-cloud 2.25 are already available from anaconda.org

**Comment by johnkerl** [comment]
@jdblischak thanks; I pushed when I'd only done half the work 😓 

Pins updated on second commit

**Review by jdblischak — APPROVED** [review]

**Comment by johnkerl** [comment]
P.S. `conda smithy rerender --commit auto` says it's up to date ✨

**Comment by johnkerl** [comment]
Still

```
Could not solve for environment specs
The following packages are incompatible
├─ aws-crt-cpp 0.26.8.* h1c89d3c_11 is requested and can be installed;
└─ libtiledbsoma 1.12.0.* h9c78a60_0 is not installable because it requires
   └─ tiledb >=2.24.0,<2.25.0a0 , which requires
      └─ aws-crt-cpp >=0.26.10,<0.26.11.0a0 , which conflicts with any installable versions previously reported.
```

Notes:

* https://github.com/conda-forge/aws-crt-cpp-feedstock
* https://anaconda.org/conda-forge/aws-crt-cpp

👀

**Comment by jdblischak** [comment]
Interesting. So from the logs, we know:

* This only affects osx-* and not linux-*
* tiledb 2.24 is built against aws-crt-cpp 0.26.10, which makes sense since that migration was merged for tiledb 2.23 (https://github.com/conda-forge/tiledb-feedstock/pull/293)
* Something is requesting aws-crt-cpp 0.26.8, which is an older version

In other words, the [new migration to aws-crt-cpp 0.26.12](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6031) is not related to this solver error. Looking at the [migration status page](https://conda-forge.org/status/migration/aws_crt_cpp02612), no PRs have been opened yet

image

**Comment by johnkerl** [comment]
@jdblischak weerd :| 

* Agreed, puzzled it's osx-only
* Thanks for confirming this isn't related to the `aws-crt-cpp` migration (although :( b/c that removes a hope for fixing this)
* In the osx build log I'm puzzled by `aws-crt-cpp 0.26.8.* h5cd1ed2_11 is requested and can be installed;` -- requested _by whom_? I'm currently trying to find out which dependency is asking for 0.26.8 ...

**Comment by johnkerl** [comment]
Oh wait -- _above_ that in the log:

```
2024-06-13T14:18:06.1806950Z WARNING: failed to get package records, retrying.  exception was: Unsatisfiable dependencies for platform osx-arm64: {MatchSpec("libtiledbsoma==1.12.0=hdc89182_0"), MatchSpec("tiledb[version='>=2.24.0,<2.25.0a0']")}
2024-06-13T14:18:06.1808150Z Encountered problems while solving:
2024-06-13T14:18:06.1808810Z   - package libtiledbsoma-1.12.0-hdc89182_0 requires tiledb >=2.24.0,<2.25.0a0, but none of the providers can be installed
```

👀 🤔

**Comment by jdblischak** [comment]
I bet it is only on osx-* because of these pins:

https://github.com/TileDB-Inc/tiledbsoma-feedstock/blob/a44ad933e5fea433a2810f3a716a7868d9984bff/recipe/meta.yaml#L94-L96

The conda-forge migrations are no longer backported to pyarrow/libarrow 12. The oldest version that still receives migrations is libarrow 13

https://github.com/conda-forge/arrow-cpp-feedstock/blob/2f976c55060605a8f6ff265d07e5ce55dad8231b/conda-forge.yml#L4-L8

Looking at the [history of PRs I've sent to arrow-cpp-feedstock](https://github.com/conda-forge/arrow-cpp-feedstock/pulls?q=is%3Apr+author%3Ajdblischak+is%3Aclosed), I see that I was the one who last migrated 12.x to aws-crt-cpp 0.26.8 (https://github.com/conda-forge/arrow-cpp-feedstock/pull/1377)

image

I can try to manually backport the latest migrations to libarrow 12.x

**Comment by johnkerl** [comment]
Garg. @jdblischak that MacOS/pyarrow bug is killing us. :(

There are several links here:

https://github.com/single-cell-data/TileDB-SOMA/pull/2692#issuecomment-2155015688

**Comment by jdblischak** [comment]
🤞 https://github.com/conda-forge/arrow-cpp-feedstock/pull/1443

**Comment by h-vetinari** [comment]
It's not a viable strategy to keep `pyarrow <13` on osx. You should figure out the issues, report it where necessary (upstream or conda-forge), and help get the issue fixed so that the cap can be removed ASAP.

Given the error originates from aws-c-common, I also doubt that it's related to the pyarrow version...🤷 You're in luck that the 12.x branch can still be rerendered and rebuilt, but this will eventually reach a point where the effort to fix becomes prohibitive (and I won't merge any other non-trivial changes - e.g. additional patch backports - other than updating the pins that come from the rerender)

File: recipe/meta.yaml [hunk 1]
```diff
@@ -91,23 +89,17 @@ outputs:
...
+        - pyarrow
       run:
         - {{ pin_compatible('numpy', lower_bound='1.16', upper_bound='1.27') }}
         - {{ pin_subpackage('libtiledbsoma', exact=True) }}
         - pandas
-        # https://github.com/single-cell-data/TileDB-SOMA/issues/1926
-        - pyarrow <13.0 # [osx]
-        - pyarrow       # [not osx]
-        - pyarrow-hotfix
+        - pyarrow
```
Comment (jdblischak) [inline recipe/meta.yaml:97]: If we are removing the fallback of pyarrow-hotfix, I think we need to set the lower bound to `>=14` (for `build`, `host`, and `run`)

```suggestion
        - pyarrow >=14
```

**Review by jdblischak — CHANGES_REQUESTED** [review]

**Comment by johnkerl** [comment]
This _solves_ OK but now I'm trying to isolate a segfault: please see
https://github.com/single-cell-data/TileDB-SOMA/pull/2692#issuecomment-2168628050

**Comment by johnkerl** [comment]
> It's not a viable strategy to keep pyarrow <13 on osx. You should figure out the issues, report it where necessary (upstream or conda-forge), and help get the issue fixed so that the cap can be removed ASAP.
> 
> Given the error originates from aws-c-common, I also doubt that it's related to the pyarrow version...🤷 You're in luck that the 12.x branch can still be rerendered and rebuilt, but this will eventually reach a point where the effort to fix becomes prohibitive (and I won't merge any other non-trivial changes - e.g. additional patch backports - other than updating the pins that come from the rerender)

@h-vetinari thank you for your feedback. I am working this as a top/unbreak priority. FWIW, some recent status is at https://github.com/single-cell-data/TileDB-SOMA/pull/2692#issuecomment-2168628050.

File: recipe/meta.yaml (hunk 1 above)
Comment (johnkerl) [reply recipe/meta.yaml]: https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/171/commits/a972b967d7e3f58f7ecaf58e1080b5a0abadb992

Thanks @jdblischak !

**Comment by johnkerl** [comment]
@h-vetinari please also see https://github.com/apache/arrow/issues/42154

**Comment by johnkerl** [comment]
Status as of commit https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/171/commits/1dce736cc700b908b2414ba5fce0694bcd8a4aa0:

* The TileDB-SOMA repo, for its wheel-build CI, simply cannot have `pyarrow>=13` on MacOS
  * Abort-trap issuepre-[2734](https://github.com/single-cell-data/TileDB-SOMA/pull/2734)
  * Segfault issue post-[2734](https://github.com/single-cell-data/TileDB-SOMA/pull/2734)
  * Looks like it will require a third-party resolution https://github.com/apache/arrow/issues/42154
* So the TileDB-SOMA repo still pins to `pyarrow<13` for MacOS, and therefore still needs `pyarrow_hotfix` for an Arrow CVE
* In this feedstock PR we have isolated `aws-sdk-cpp`
  * The test on this PR proves we're immune to the abort/segfault issues ☝️ 
  * But the repo source code still does `import pyarrow_hotfix`
  * So we need both `pyarrow>=14` and `import pyarrow_hotfix` for now
* https://github.com/single-cell-data/TileDB-SOMA/issues/1926 tracks all this

**Comment by johnkerl** [comment]
numpy 2 was released yesterday

**Comment by johnkerl** [comment]
All ☘️ now -- closing this pre-check PR per our [established procedure](https://github.com/single-cell-data/TileDB-SOMA/wiki/Branches-and-releases).

**Pull request closed without merging**